contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,222 @@
+# DaemonSet Updates
+
+**Author**: @madhusudancs, @lukaszo, @janetkuo
+
+**Status**: Proposal
+
+## Abstract
+
+A proposal for adding the update feature to `DaemonSet`. This feature will be
+implemented on server side (in `DaemonSet` API). 
+
+Users already can update a `DaemonSet` today (Kubernetes release 1.5), which will
+not cause changes to its subsequent pods, until those pods are killed. In this
+proposal, we plan to add a "RollingUpdate" strategy which allows DaemonSet to
+downstream its changes to pods. 
+
+## Requirements
+
+In this proposal, we design DaemonSet updates based on the following requirements:
+
+- Users can have a controlled number of in-flight updates of a DaemonSet (e.g.
+  DaemonSet only updates one pod at a time)
+- Users can monitor the status of a DaemonSet update (e.g. the number of pods
+  that are updated and healthy)
+- A broken DaemonSet update should not continue 
+- Users should be able to update a DaemonSet, even during an ongoing DaemonSet
+  upgrade (e.g. update the DaemonSet to fix a broken DaemonSet update)
+- Users should be able to view the history of previous DaemonSet updates 
+- Users can figure out the revision of a DaemonSet's pod (e.g. which version is
+  this DaemonSet pod?)
+
+Here are some potential requirements that haven't been covered by this proposal:
+
+- Users can have rate-limited DaemonSet updates (e.g. only upgrade 10% per hour)
+- DaemonSet should provide at-most-one guarantee per node (i.e. at most one pod
```

Comment:
In current implementation, DS controller lists pods from a pod cache store. It's possible that DS controller hasn't observed a daemon pod which is just created on a node, and then the controller will try to create another pod on that same node. I haven't seen this bug happen so I just imagine it would happen, especially when people start rolling update their DaemonSets. 

One way to prevent this is to use pod name as a lock, like what StatefulSet does. We can use node name as pod suffix (instead of a random string), so that it's not possible to create two daemon pods (from the same DS) on the same node.